dipy/reconst/ivim.py

```diff
@@ -0,0 +1,383 @@
+        if x0 is None:
+            x0 = np.ones(
+                (data.shape[:-1] + (4,))) * [1.0, 0.10, 0.001, 0.0009]
+        else:
+            if x0.shape != (data.shape[0], 4):
+                raise ValueError(
+                    "Guess params should be of shape (num_voxels,4)")
+        data_in_mask = np.maximum(data_in_mask, min_signal)
+        x0 = np.reshape(x0[mask], (-1, x0.shape[-1]))
+
+        if fit_method == "one_stage":
+            params_in_mask = one_stage(data_in_mask, self.gtab, x0,
+                                       jac, bounds, tol,
+                                       routine,
+                                       algorithm, gtol, ftol, eps, maxiter)
+        elif fit_method == "two_stage":
+            params_in_mask = two_stage(data_in_mask, self.gtab, x0,
+                                       self.split_b, jac, bounds, tol,
+                                       routine,
+                                       algorithm, gtol, ftol, eps, maxiter)
+        else:
+            raise ValueError("""Fit method must be either
+                             'one_stage' or 'two_stage'""")
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            ivim_params = params_in_mask.reshape(out_shape)
+        else:
+            ivim_params = np.zeros(data.shape[:-1] + (4,))
+            ivim_params[mask, :] = params_in_mask
+
+        return IvimFit(self, ivim_params)
+
+    def predict(self, ivim_params):
+        """
+        Predict a signal for this IvimModel class instance given parameters.
+
+        Parameters
+        ----------
+        ivim_params : ndarray
+            The last dimension should have 4 parameters: S0, f, D_star and D
+        """
+
+        return ivim_function(ivim_params, self.gtab.bvals)
+
+
+class IvimFit(object):
+
+    def __init__(self, model, model_params):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            The model parameters are S0, f, D_star, D
+        """
+        self.model = model
+        self.model_params = model_params
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+    def predict(self, gtab, step=None, S0=1.0):
+        r"""
+        Given a model fit, predict the signal.
+
+        Parameters
+        ----------
+        gtab : a GradientTable class instance
+            This encodes the directions for which a prediction is made
+
+        Notes
+        -----
+        The predicted signal is given by:
+
+        .. math ::
+
```

Comment:
This should be filled in. I would actually be happy to have these equations appear earlier in the documentation of the `IvimModel` class as well.